sdk/java/pulumi/src/main/java/com/pulumi/serialization/internal/Converter.java

```diff
@@ -530,52 +553,68 @@ private void tryWriteJson(String context, JsonWriter jsonWriter, @Nullable Objec
         ));
     }
 
-    private <T> T tryEnsureType(String context, Object value, TypeShape<T> targetType) {
-        if (targetType.getType().isInstance(value)
+    private <T> boolean canBeCasted(Object value, TypeShape<T> targetType) {
+        return targetType.getType().isInstance(value)
                 || (boolean.class.isAssignableFrom(targetType.getType()) && value instanceof Boolean)
                 || (double.class.isAssignableFrom(targetType.getType()) && value instanceof Double)
                 || (int.class.isAssignableFrom(targetType.getType()) && value instanceof Integer)
-        ) {
+                || (int.class.isAssignableFrom(targetType.getType()) && value instanceof Double)
+                || (Integer.class.isAssignableFrom(targetType.getType()) && value instanceof Double);
+    }
+
+    @Nullable
+    private <T> T tryEnsureType(String context, Object value, TypeShape<T> targetType, T default_) {
+        if (canBeCasted(value, targetType)) {
             //noinspection unchecked
             return (T) value;
         } else {
-            throw new UnsupportedOperationException(String.format(
+            this.log.warn(String.format(
                     "%s; Expected '%s' but got '%s' while deserializing.",
                     context, targetType.getTypeName(), value.getClass().getTypeName()
             ));
+            return defaultValue(targetType, default_);
         }
     }
 
+    @Nullable
     private Either<Object, Object> tryConvertOneOf(String context, Object value, TypeShape<?> targetType) {
         var leftType = targetType.getParameter(0)
                 .orElseThrow(() -> new IllegalStateException("Expected a left parameter type for the Either, got none"));
         var rightType = targetType.getParameter(1)
-                .orElseThrow(() -> new IllegalStateException("Expected a left parameter type for the Either, got none"));
+                .orElseThrow(() -> new IllegalStateException("Expected a right parameter type for the Either, got none"));
 
         try {
-            return Either.ofLeft(
-                    tryConvertObjectInner(
-                            String.format("%s.left", context),
-                            value,
-                            leftType
-                    )
-            );
-        } catch (Exception leftException) {
-            try {
+            if (canBeCasted(value, leftType)) {
```

Comment:
https://grammarist.com/usage/casted/#:~:text=The%20verb%20cast%20is%20conventionally,to%20cast%20in%20modern%20English. perhaps `canBeCast`.

Reply:
thank you